File: src/liveplugin/pluginrunner/ScalaPluginRunner.java
package liveplugin.pluginrunner;

import com.intellij.openapi.application.PathManager;
import com.intellij.openapi.util.io.FileUtil;
import com.intellij.util.Function;
import com.intellij.util.PathUtil;
import liveplugin.MyFileUtil;
import scala.Some;
import scala.package$;
import scala.tools.nsc.Settings;
import scala.tools.nsc.interpreter.IMain;
import scala.tools.nsc.interpreter.Results;
import scala.tools.nsc.settings.MutableSettings;
import scala.xml.Null;

import java.io.File;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import static com.intellij.openapi.application.PathManager.getLibPath;
import static com.intellij.openapi.application.PathManager.getPluginsPath;
import static com.intellij.openapi.util.text.StringUtil.join;
import static com.intellij.util.containers.ContainerUtil.flatten;
import static com.intellij.util.containers.ContainerUtil.map;
import static com.intellij.util.containers.ContainerUtilRt.newArrayList;
import static java.io.File.pathSeparator;
import static java.util.Arrays.asList;
import static liveplugin.MyFileUtil.*;
import static liveplugin.pluginrunner.PluginRunner.ClasspathAddition.*;

/**
 * This class should not be loaded unless scala libs are on classpath.
 */
public class ScalaPluginRunner implements PluginRunner {
	private static final String SCALA_DEPENDS_ON_PLUGIN_KEYWORD = "// " + DEPENDS_ON_PLUGIN_KEYWORD;
	public static final String MAIN_SCRIPT = "plugin.scala";
	private static final String SCALA_ADD_TO_CLASSPATH_KEYWORD = "// " + ADD_TO_CLASSPATH_KEYWORD;
	private static final StringWriter interpreterOutput = new StringWriter();
	private static final Object interpreterLock = new Object();

	private final ErrorReporter errorReporter;
	private final Map<String, String> environment;


	public ScalaPluginRunner(ErrorReporter errorReporter, Map<String, String> environment) {
		this.errorReporter = errorReporter;
		this.environment = environment;
	}

	private static IMain initInterpreter(String interpreterClasspath, ClassLoader parentClassLoader) throws ClassNotFoundException {
		Settings settings = new Settings();
		MutableSettings.PathSetting bootClasspath = (MutableSettings.PathSetting) settings.bootclasspath();
		bootClasspath.append(interpreterClasspath);

		settings.explicitParentLoader_$eq(new Some<>(parentClassLoader));

		((MutableSettings.BooleanSetting) settings.usejavacp()).tryToSetFromPropertyValue("true");

		return new IMain(settings, new PrintWriter(interpreterOutput));
	}

	private static String createInterpreterClasspath(List<String> additionalPaths) throws ClassNotFoundException {
		Function<File, String> toAbsolutePath = it -> it.getAbsolutePath();
		Function<File, Collection<File>> findPluginJars = pluginPath -> {
			if (pluginPath.isFile()) {
				return newArrayList(pluginPath);
			} else {
				return newArrayList(withDefault(new File[0], new File(pluginPath, "lib").listFiles((file, fileName) -> fileName.endsWith(".jar") || fileName.endsWith(".zip"))));
			}
		};

		String compilerPath = PathUtil.getJarPathForClass(Class.forName("scala.tools.nsc.Interpreter"));
		String scalaLibPath = PathUtil.getJarPathForClass(Class.forName("scala.Some"));
		String intellijLibPath = join(map(withDefault(new File[0], new File(getLibPath()).listFiles()), toAbsolutePath), pathSeparator);
		String allNonCorePluginsPath = join(map(flatten(map(withDefault(new File[0], new File(getPluginsPath()).listFiles()), findPluginJars)), toAbsolutePath), pathSeparator);
		String livePluginPath = PathManager.getResourceRoot(ScalaPluginRunner.class, "/liveplugin/"); // this is only useful when running liveplugin from IDE (it's not packed into jar)
		return join(asList(compilerPath, scalaLibPath, livePluginPath, intellijLibPath, allNonCorePluginsPath), pathSeparator) +
				pathSeparator + join(additionalPaths, pathSeparator);
	}

	private static <T> T withDefault(T defaultValue, T value) {
		return value == null ? defaultValue : value;
	}

	@Override public boolean canRunPlugin(String pathToPluginFolder) {
		return findScriptFileIn(pathToPluginFolder, MAIN_SCRIPT) != null;
	}

	@Override public void runPlugin(String pathToPluginFolder, final String pluginId,
	                                Map<String, ?> binding, Function<Runnable, Void> runOnEDTCallback) {
		final File scriptFile = MyFileUtil.findScriptFileIn(pathToPluginFolder, ScalaPluginRunner.MAIN_SCRIPT);
		assert scriptFile != null;

		final IMain interpreter;
		synchronized (interpreterLock) {
			try {
				environment.put("PLUGIN_PATH", pathToPluginFolder);

				List<String> dependentPlugins = findPluginDependencies(readLines(asUrl(scriptFile)), SCALA_DEPENDS_ON_PLUGIN_KEYWORD);
				List<String> additionalPaths = findClasspathAdditions(readLines(asUrl(scriptFile)), SCALA_ADD_TO_CLASSPATH_KEYWORD, environment, path -> {
					errorReporter.addLoadingError(pluginId, "Couldn't find dependency '" + path + "'");
					return null;
				});
				String classpath = createInterpreterClasspath(additionalPaths);
				ClassLoader parentClassLoader = createParentClassLoader(dependentPlugins, pluginId, errorReporter);
				interpreter = initInterpreter(classpath, parentClassLoader);

			} catch (Exception | LinkageError e) {
				errorReporter.addLoadingError("Failed to init scala interpreter", e);
				return;
			}

			interpreterOutput.getBuffer().delete(0, interpreterOutput.getBuffer().length());
			for (Map.Entry<String, ?> entry : binding.entrySet()) {
				String key = entry.getKey();
				Object value = entry.getValue();
				String valueClassName = value == null ? Null.class.getCanonicalName() : value.getClass().getCanonicalName();
				interpreter.bind(key, valueClassName, value, package$.MODULE$.List().empty());
			}
		}

		runOnEDTCallback.fun(() -> {
			synchronized (interpreterLock) {
				Results.Result result;
				try {
					result = interpreter.interpret(FileUtil.loadFile(scriptFile));
				} catch (LinkageError e) {
					errorReporter.addLoadingError(pluginId, "Error linking script file: " + scriptFile);
					return;
				} catch (Exception e) {
					errorReporter.addLoadingError(pluginId, "Error reading script file: " + scriptFile);
					return;
				}

				if (!(result instanceof Results.Success$)) {
					errorReporter.addRunningError(pluginId, interpreterOutput.toString());
				}
			}
		});
	}

	@Override public String scriptName() {
		return MAIN_SCRIPT;
	}
}

File: src/main/liveplugin/pluginrunner/ScalaPluginRunner.kt
package liveplugin.pluginrunner

import com.intellij.openapi.application.PathManager
import com.intellij.openapi.application.PathManager.getLibPath
import com.intellij.openapi.application.PathManager.getPluginsPath
import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.util.text.StringUtil.join
import com.intellij.util.Function
import com.intellij.util.PathUtil
import com.intellij.util.containers.ContainerUtil.flatten
import com.intellij.util.containers.ContainerUtil.map
import liveplugin.MyFileUtil
import liveplugin.MyFileUtil.*
import liveplugin.pluginrunner.PluginRunner.ClasspathAddition.createParentClassLoader
import liveplugin.pluginrunner.PluginRunner.ClasspathAddition.findClasspathAdditions
import liveplugin.pluginrunner.PluginRunner.ClasspathAddition.findPluginDependencies
import scala.Some
import scala.`package$`
import scala.tools.nsc.Settings
import scala.tools.nsc.interpreter.IMain
import scala.tools.nsc.interpreter.Results
import scala.tools.nsc.settings.MutableSettings
import scala.xml.Null
import java.io.File
import java.io.File.pathSeparator
import java.io.PrintWriter
import java.io.StringWriter
import java.util.Arrays.asList

/**
 * This class should not be loaded unless scala libs are on classpath.
 */
class ScalaPluginRunner(private val errorReporter: ErrorReporter, private val environment: MutableMap<String, String>): PluginRunner {

    override fun canRunPlugin(pathToPluginFolder: String): Boolean {
        return findScriptFileIn(pathToPluginFolder, mainScript) != null
    }

    override fun runPlugin(pathToPluginFolder: String, pluginId: String,
                           binding: Map<String, *>, runOnEDTCallback: Function<Runnable, Void>) {
        val scriptFile = MyFileUtil.findScriptFileIn(pathToPluginFolder, ScalaPluginRunner.mainScript)!!

        var interpreter: IMain? = null
        synchronized(interpreterLock) {
            try {
                environment.put("PLUGIN_PATH", pathToPluginFolder)

                val dependentPlugins = findPluginDependencies(readLines(asUrl(scriptFile)), scalaDependsOnPluginKeyword)
                val additionalPaths = findClasspathAdditions(readLines(asUrl(scriptFile)), scalaAddToClasspathKeyword, environment, onError = { path ->
                    errorReporter.addLoadingError(pluginId, "Couldn't find dependency '$path'")
                })
                val classpath = createInterpreterClasspath(additionalPaths)
                val parentClassLoader = createParentClassLoader(dependentPlugins, pluginId, errorReporter)
                interpreter = initInterpreter(classpath, parentClassLoader)

            } catch (e: Exception) {
                errorReporter.addLoadingError("Failed to init scala interpreter", e)
                return
            } catch (e: LinkageError) {
                errorReporter.addLoadingError("Failed to init scala interpreter", e)
                return
            }

            interpreterOutput.buffer.delete(0, interpreterOutput.buffer.length)
            for ((key, value) in binding) {
                val valueClassName = if (value == null) Null::class.java.canonicalName else value.javaClass.canonicalName
                interpreter!!.bind(key, valueClassName, value, `package$`.`MODULE$`.List().empty())
            }
        }

        runOnEDTCallback.`fun`(Runnable {
            synchronized(interpreterLock) {
                val result: Results.Result
                try {
                    result = interpreter!!.interpret(FileUtil.loadFile(scriptFile))
                    if (result !is Results.`Success$`) {
                        errorReporter.addRunningError(pluginId, interpreterOutput.toString())
                    }
                } catch (e: LinkageError) {
                    errorReporter.addLoadingError(pluginId, "Error linking script file: " + scriptFile)
                } catch (e: Exception) {
                    errorReporter.addLoadingError(pluginId, "Error reading script file: " + scriptFile)
                }
            }
        })
    }

    override fun scriptName(): String {
        return mainScript
    }

    companion object {
        private val scalaDependsOnPluginKeyword = "// " + PluginRunner.dependsOnPluginKeyword
        @JvmField val mainScript = "plugin.scala"
        private val scalaAddToClasspathKeyword = "// " + PluginRunner.addToClasspathKeyword
        private val interpreterOutput = StringWriter()
        private val interpreterLock = Any()

        @Throws(ClassNotFoundException::class)
        private fun initInterpreter(interpreterClasspath: String, parentClassLoader: ClassLoader): IMain {
            val settings = Settings()
            val bootClasspath = settings.bootclasspath() as MutableSettings.PathSetting
            bootClasspath.append(interpreterClasspath)

            settings.`explicitParentLoader_$eq`(Some(parentClassLoader))

            (settings.usejavacp() as MutableSettings.BooleanSetting).tryToSetFromPropertyValue("true")

            return IMain(settings, PrintWriter(interpreterOutput))
        }

        @Throws(ClassNotFoundException::class)
        private fun createInterpreterClasspath(additionalPaths: List<String>): String {
            val findPluginJars = { pluginPath: File ->
                if (pluginPath.isFile) {
                    listOf(pluginPath)
                } else {
                    File(pluginPath, "lib").listFiles({ _, fileName -> fileName.endsWith(".jar") || fileName.endsWith(".zip") })?.toList() ?: emptyList()
                }
            }

            val compilerPath = PathUtil.getJarPathForClass(Class.forName("scala.tools.nsc.Interpreter"))
            val scalaLibPath = PathUtil.getJarPathForClass(Class.forName("scala.Some"))
            val intellijLibPath = join(map(withDefault(arrayOfNulls<File>(0), File(getLibPath()).listFiles())) { it.absolutePath }, pathSeparator)
            val allNonCorePluginsPath = join(map(flatten(map(withDefault(arrayOfNulls<File>(0), File(getPluginsPath()).listFiles()), findPluginJars)), { it.absolutePath }), pathSeparator)
            val livePluginPath = PathManager.getResourceRoot(ScalaPluginRunner::class.java, "/liveplugin/") // this is only useful when running liveplugin from IDE (it's not packed into jar)
            return join(asList<String>(compilerPath, scalaLibPath, livePluginPath, intellijLibPath, allNonCorePluginsPath), pathSeparator) +
                pathSeparator + join(additionalPaths, pathSeparator)
        }

        private fun <T> withDefault(defaultValue: T, value: T?): T {
            return value ?: defaultValue
        }
    }
}
